### PR TITLE
nvfuser update

### PR DESCRIPTION
**Pull request by jjsjann123**
Syncing nvfuser code base from devel branch, Listing a few of our development since last sync:

- Extends support to normalization and reduction kernels.
- Multiple kernel launch for single `CudaFusionGroup`. Hierarchical caching system has been updated to cache graph segmentation.
- profile_ivalue is enabled to convert dynamic scalar into compile time constants, which are required by the codegen. (e.g. reduction axes).

To keep this PR simple and relatively review-free. We stripped most external changes and submitted them as separate PRs, so this gigantic PR is easier to handle.

internal updates are files located in:
1. updates in nvfuser codegen `torch/csrc/jit/coddgen/cuda`
2. added nvfuser specific benchmarks `benchmarks/cpp/nvfuser`
3. nvfuser jit cpp tests `test/cpp/jit/test_gpu.cpp` `test/cpp/jit/test_gpu_shift.cpp` `test/cpp/jit/test_gpu_validator.h`

updates affecting integration:

1. profile_ivalue enabled for nvfuser. related changes are in `torch/csrc/jit/runtime/*`, 
2. exposed a few more symbols `aten/src/ATen/core/*` used by codegen

### CODE REVIEW

**Comment by jjsjann123** [comment]
Pointed out by Nick that we are also touching some build files (for the added benchmark folder). Adding @malfet to reviewer list.

**Review by Krovatkin — APPROVED** [review]
:shipit:

**Review by csarofeen — APPROVED** [review]

**Comment by jjsjann123** [comment]
Pinging @malfet on build file changes. :pray:

**Comment by jjsjann123** [comment]
Pinging on the progress. Wondering if there's anything coming up in internal CI?

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 1]
```diff
@@ -201,7 +205,8 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
 }
 
 bool needsProfiledInputs(Node* n) {
-  if (tensorexpr::isSupported(n)) {
+  if (tensorexpr::isSupported(n) ||
+      (RegisterCudaFuseGraph::isRegistered() && fuser::cuda::canFuseNode(n))) {
```
Comment (malfet) [inline torch/csrc/jit/runtime/profiling_record.cpp:209]: Calls to `fuser::` needs to be guarded by `!C10_MOBILE`
```suggestion
  if (tensorexpr::isSupported(n) ||
#ifndef C10_MOBILE
      (RegisterCudaFuseGraph::isRegistered() && fuser::cuda::canFuseNode(n))
#else
      false
#endif
     ) {
```

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 2]
```diff
@@ -232,7 +237,8 @@ bool needsProfiledInputs(Node* n) {
 }
 
 bool needsProfiledOutput(Node* n) {
-  if (tensorexpr::isSupported(n)) {
+  if (tensorexpr::isSupported(n) ||
+      (RegisterCudaFuseGraph::isRegistered() && fuser::cuda::canFuseNode(n))) {
```
Comment (malfet) [inline torch/csrc/jit/runtime/profiling_record.cpp:241]: Same as above
```suggestion
#ifndef C10_MOBILE
      (RegisterCudaFuseGraph::isRegistered() && fuser::cuda::canFuseNode(n))
#else
      false
#endif
     ) {
```

**Review by malfet — COMMENTED** [review]
Applied suggested changes to internal copy of diff to solve build problems

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 3]
```diff
@@ -668,6 +668,11 @@ const ExecutionPlan& ProfilingGraphExecutorImpl::getOptimizedPlanFor(
     // before any other pass that could insert `prim::iprofile_value` node on
     // `aten::_grad_sum_to_size` input.
     InsertProfileNodesForSpecializeAutogradZero(pr_.get());
+    // `InsertProfileNodesForCUDAFuser` inserts profile node for non-tensor
+    // value
+    if (RegisterCudaFuseGraph::isRegistered()) {
+      torch::jit::fuser::cuda::InsertProfileNodesForCUDAFuser(pr_.get());
+    }
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:676]: Updated

**Comment by jjsjann123** [comment]
mypy is complaining on files not changed in this PR. Seems to be a false alarm.

**Comment by seemethere** [comment]
Can we not have giant PRs like this? This makes it extremely hard to merge as well as making it extremely hard to debug where issues might've occured

**Comment by seemethere** [comment]
The reason why this was reverted is that it caused builds to fail, see:
* (on this pr): https://github.com/pytorch/pytorch/runs/3608523266?check_suite_focus=true
* (upstream builds): https://hud.pytorch.org/commit/cfaecaf40bd6cabd3f4e0ef0d8c7252655349b61

**Comment by jjsjann123** [comment]
:cry:

Huge PR is indeed tricky to handle. I should have done a better job at
going through CI logs, but I don't see failures that are related to changes
in this PR. Any more pointers? @seemethere
Meanwhile, I'll reopen a PR to revert the revert, just to kick off CI again
and hoping that would give me more ideas.

Eli Uriegas ***@***.***> 于2021年9月15日周三 下午5:49写道：

**Comment by seemethere** [comment]
So for a change as big as this one I'd recommend adding two different labels:
* ci/master (which should enable circleci jobs that run on master)
* ciflow/all (which should enable all GHA jobs that run on master, including periodic jobs)

**Pull request closed without merging**